synapse/tests/test_lib_crypto_vault.py

```diff
@@ -0,0 +1,199 @@
+            ncert = newvault.addUserAuth(auth)
+            self.isinstance(ncert, s_vault.Cert)
+            ncert = newvault.getUserCert('bobgrey@vertex.link')
+            self.eq(ncert.iden(), cert.iden())
+            # This cert is not signed legitimately according to newvault
+            newvault.isValidCert(ncert)
+
+            # We can manipulate CA data
+            nroot = newvault.genRootCert()
+            self.nn(nroot)
+            self.len(1, vault.getRootCerts())
+            self.none(vault.addRootCert(nroot))
+            self.len(2, vault.getRootCerts())
+            # Since we added newvault's root CA to vault,
+            # it can validate certs from newvault
+            ncert = newvault.genUserCert('hehe@haha.com')
+            self.true(vault.isValidCert(ncert))
+            # But deleted certs don't go away from the DB
+            self.nn(vault.getCert(nroot.iden()))
+            # And we can delete root certs
+            self.none(vault.delRootCert(nroot))
+            self.len(1, vault.getRootCerts())
+            # vault can no longer validate data from newvault
+            self.false(vault.isValidCert(ncert))
+
+            # Here be dragons!
+            # XXX Ensure we know why/what we're doing with this API
+            # Ensure we can make a new RSA key directly
+            rkey = vault.genRsaKey()
+            self.isinstance(rkey, s_rsa.PriKey)
+            # We can set that key to a given user
+
+            # Tear down our vaults
+            newvault.fini()
+            vault.fini()
+
+    def test_lib_crypto_vault_shared_interproc(self):
+        with self.getTestDir() as dirn:
+            path = os.path.join(dirn, 'vault.lmdb')
+
+            evt1 = multiprocessing.Event()
+            evt1.clear()
+            evt2 = multiprocessing.Event()
+            evt2.clear()
+
+            user = 'pennywise@vertex.link'
+
+            # Fire a process which makes a user cert in the vault after one
+            # proc already obtains the advisory lock.  This will be blocked
+            # by the advisory file lock.
+            with s_vault.shared(path) as vault:
+                # thr = doStuff(path, 'pennywise@local.link')
+                proc = multiprocessing.Process(target=addUserToVault,
+                                               args=(evt1, evt2, path, user))
+                proc.start()
+
+                cert = vault.genUserCert('bobgrey@vertex.link')
+                self.nn(cert)
+                evt1.set()
+                self.none(vault.getUserCert(user))
+
+            evt2.wait(10)
+            proc.join(10)
+
+            # Ensure that the user is present in the vault now after the
+            # main process has released its lock
+            with s_vault.shared(path) as vault:
+                self.nn(vault.getUserCert(user))
+
+    def test_lib_crypto_vault_shared_intraproc(self):
+        self.skipTest('Skipped pending shared() locked addition')
```

Comment:
this test fails periodically with lmdb race conditions (see earlier comment about locking)